common/flatpak-run.c

```diff
@@ -3092,8 +3092,94 @@ flatpak_run_add_app_info_args (GPtrArray      *argv_array,
+{
+  const char *runtime_files_path = NULL;
+  g_autofree char *deployed_target = NULL;
+
+  /* Manually check whether the symlink would point to a resolvable path inside the sandbox. */
+  runtime_files_path = flatpak_file_get_path_cached (runtime_files);
+  if (target[0] == '/')
+    {
+      /* Symlink points to an absolute path */
+      guint base_index = (g_strcmp0 (target, "/usr") == 0) ? 4 : 1;
+      deployed_target = g_strdup_printf ("%s/%s", runtime_files_path, &target[base_index]);
+    }
+  else
+    {
+      g_autofree char *symlink_subdir = NULL;
+      symlink_subdir = g_path_get_dirname (symlink);
+
+      if (!g_str_has_prefix (target, "../"))
+        {
+          /* Symlink points to a file relative to its base directory */
+          deployed_target = g_strdup_printf ("%s/%s/%s", runtime_files_path, symlink_subdir, target);
+        }
+      else
+        {
+          /* Symlink points to a file relative to and ancestor of its base directory */
+          g_autofree char *target_subdir = NULL;
+          g_autofree char *target_root_subdir = NULL;
+          g_autofree char *target_basename = NULL;
+          g_auto(GStrv) symlink_subdirs = NULL;
+          g_auto(GStrv) target_subdirs = NULL;
+          const char *target_subdir_last_part = NULL;
+          guint n_symlink_path_elems;
+          guint n_target_path_elems;
+
+          target_subdir = g_path_get_dirname (target);
+
+          symlink_subdirs = g_strsplit (symlink_subdir, "/", -1);
+          n_symlink_path_elems = g_strv_length (symlink_subdirs);
+
+          target_subdirs = g_strsplit (target_subdir, "../", -1);
+          n_target_path_elems = g_strv_length (target_subdirs);
+
+          target_subdir_last_part = target_subdirs[n_target_path_elems - 1];
+
+          /* We need to check whether the relative symlink is referencing '/usr' and remove that
+           * part from the path if that's the case, as it won't be available in the deply directory.
+           */
+          if (n_target_path_elems >= n_symlink_path_elems && g_str_has_prefix (target_subdir_last_part, "usr/"))
+            target_root_subdir = g_strdup (&target_subdir_last_part[4]);
+          else
+            target_root_subdir = g_strdup_printf ("%s/%s", symlink_subdir, target_subdir);
+
+          target_basename = g_path_get_basename (target);
+          deployed_target = g_strdup_printf ("%s/%s/%s", runtime_files_path, target_root_subdir, target_basename);
+        }
+    }
+
+  return g_file_test (deployed_target, G_FILE_TEST_EXISTS);
+}
+
+static void
+add_unmonitored_config_file_args (GFile *runtime_files, const char *config_file_path, GPtrArray *argv_array)
+{
+  char sysmlink_target[PATH_MAX + 1];
+  ssize_t symlink_size;
+
+  if (!g_file_test (config_file_path, G_FILE_TEST_EXISTS))
+    return;
+
+  symlink_size = readlink (config_file_path, sysmlink_target, sizeof (sysmlink_target) - 1);
+  if (symlink_size > 0)
+    sysmlink_target[symlink_size] = 0;
```

Comment:
For non-symlinks, (symlink_size <= 0) we do nothing, which passes on uninitialzed datat to syslink_target_is_deployed below. That can't be right.